### DISCUSSION TITLE
Multi-animal tracking pipeline

### DISCUSSION THREAD

**Discussion by glopesdev** [SainsburyWellcomeCentre/aeon_experiments#141] [General]
Checklist:

- [x] Repository for storing pipeline code (new branch on [aeon_analysis](https://github.com/SainsburyWellcomeCentre/aeon_analysis))
  - [aeon_analysis/social/aeon_sleap.ipynb ](https://github.com/SainsburyWellcomeCentre/aeon_analysis/blob/multianimal_tracking/aeon_analysis/social/aeon_sleap.ipynb) on the [multianimal_tracking branch](https://github.com/SainsburyWellcomeCentre/aeon_analysis/tree/multianimal_tracking)
- [ ] Resolve incompatible dependencies between SLEAP and aeon_mecha (we can open issues on aeon_mecha)
- [ ] Decide which dataset to use for generating training data (the one in B2 [no single animal] and an older one in 164 [single animal])
  - [ ] Each animal running individual
  - [ ] Both animals running together
- [x] Potentially check the newer B2 dataset to see if there are annotations or periods where animals are separated. 
   - Could not find individual session for mouse with tail painted. Opened #162.

**Comment by lochhh** [comment]
Current workflow:
1. create slp label dataset with skeleton containing 1 bodypart (centroid) and labeled instances
   1. given single-animal video (animal with known ID), extract centroid pos (x,y) and ID of the animal
   2. given multi-animal video, find frames where animals cross each other, manually label centroids and identities in sleap
2. train centroid + confmap with multiclass with slp datasets from 2 animals running individually
3. infer on multi-animal video

Questions:
1. Are there added benefits from training frames containing multiple animals when they are apart vs training frames containing single animals?
3. Are there any disadvantages to using a simple skeleton (e.g., with one body part)? 
4. Network trained on progressively more animals? 
5. Train on identity of the animal vs train on feature (e.g., painted tail vs not painted)?
6. Recommended training parameters - ID model - does the loss scale by no. of body parts? 
7. PredictedInstance.tracking_score=0.00

**Reply by bruno-f-cruz** [reply]
After talking to Talmo we decided to:

- Try to sample more uniformly from the arena since before we noticed that the feature extractor was biased towards the edges. E.g.:
<img width="490" alt="MicrosoftTeams-image" src="https://user-images.githubusercontent.com/14955489/220161499-7d129c6e-30d0-42f7-aeea-8a1ca7a21035.png">

- Set the bounding box of the centroid head to the confmap to 128x128.
- Set the weight of the Id head to 0.5 or 1.0 and pose to 0.01

**Reply by lochhh** [reply]
**Datasets used**
```python
AEON2 = {
    "root": "/ceph/aeon/aeon/data/raw/AEON2/experiment0.2/",
    "subjects": {
        "BAA-1101818": { #tail paint
            "start": pd.Timestamp("2022-06-23 08:39:04.261089801"),
            "end": pd.Timestamp("2022-06-23 11:14:46.121759892"),
        },
        "BAA-1101819": { # no tail paint
            "start": pd.Timestamp("2022-06-21 13:28:10.593659878"),
            "end": pd.Timestamp("2022-06-21 16:34:29.241280079"),
        },
    },
}

AEON3A = { # blurry frames
    "root": "Z:/aeon/data/raw/AEON3/presocial0.1",
    "subjects": {
        "AEON3_NTP": {
            "start": pd.Timestamp("2023-03-03 16:40:00"),
            "end": pd.Timestamp("2023-03-03 16:55:00"),
        },
        "AEON3_TP": {
            "start": pd.Timestamp("2023-03-03 17:01:00"),
            "end": pd.Timestamp("2023-03-03 17:22:00"),
        },
    },
}

AEON3B = { 
    "root": "Z:/aeon/data/raw/AEON3/presocial0.1",
    "subjects": {
        "AEON3B_NTP": {
            "start": pd.Timestamp("2023-03-16 15:05:00"),
            "end": pd.Timestamp("2023-03-16 15:44:00"),
        },
        "AEON3B_TP": {
            "start": pd.Timestamp("2023-03-16 16:00:00"),
            "end": pd.Timestamp("2023-03-16 16:36:00"),
        },
    },
}
```
**Frame sampling method** 
randomly sample 3 frames from each of the 50x50 bins  (can be fewer when the bin has < 3 frames) 

**Stats on test set with frames containing single animals**
| Train params & stats/Dataset | AEON2  | AEON2  | AEON2  | AEON2  | AEON3A | AEON3A | AEON3B | AEON3B |
|------------------------------|--------|--------|--------|--------|--------|--------|--------|--------|
| Bounding box                 | 128    | 128    | 128    | 128    | 128    | 128    | 128    | 128    |
| Batch size                   | 4      | 8      | 8      | 8      | 8      | 8      | 8      | 8      |
| Confmap head                 |        |        |        |        |        |        |        |        |
| Sigma                        | 1.5    | 1.5    | 1.5    | 1.5    | 2.5    | 2.5    | 2.5    | 1.5    |
| Output stride                | 2      | 2      | 2      | 2      | 4      | 4      | 4      | 2      |
| Loss weight                  | 0.001  | 0.001  | 0.001  | 1.0    | 1.0    | 1.0    | 1.0    | 1.0    |
| ID head                      |                                                                       |
| Output stride                | 2      | 2      | 2      | 2      | 4      | 4      | 4      | 2      |
| Loss weight                  | 0.5    | 1.0    | 2.0    | 0.01   | 0.01   | 0.1    | 0.1    | 0.1    |
| Tracks identified            | 821    | 816    | 663    | 953    | 953    | 953    | 965    | 965    |    
| Tracks correctly identified: | 773    | 749    | 624    | 846    | 870    | 868    | 956    | 959    |      
| Total tracks                 | 953    | 953    | 953    | 953    | 953    | 953    | 965    | 965    |
| ID accuracy                  | 0.8111 | 0.7859 | 0.6548 | 0.8877 | 0.9129 | 0.9108 | 0.9907 | 0.9938 |

**Reply by lochhh** [reply]
To do:
1. Check centroid model alone, retrain if needed

- [x] Validate that the [centroid model](https://sleap.ai/develop/api/sleap.nn.config.model.html?highlight=centroid#sleap.nn.config.model.CentroidsHeadConfig) on single animal videos. This should result in a high percentage of detections
  - Centroid model is detecting 953/953 instances in 8 single-animal videos
- [x] Validate the same model in multi-animal video. The model should detect 2 instances per frame when animals are apart and 0-2 when fused. 
  - [SLEAP](https://github.com/talmolab/sleap/issues/885) requires both models when predicting on non-ground-truth data
  - As a workaround, generated "ground-truth data" using a TopDownPredictor.
  - Centroid model is detecting 11309/12000 instances in a multi-animal video
<img width="84" alt="image" src="https://user-images.githubusercontent.com/14955489/223506116-a8e4a8ae-699c-4331-9eec-6ca46780ba91.png"> <img width="112" alt="image" src="https://user-images.githubusercontent.com/14955489/223506669-18863ca0-971d-410c-b53f-eb02013b655f.png">

**Comment by bruno-f-cruz** [comment]
Current issues:

- [ ] Network output format from export functionality in SLEAP has greatly changed. We need to change the Bonsai package to accommodate this.
- [ ] We need a way to look at the raw output from the network. Looking at the output in Bonsai something weird is going on. The centroid confidence is always very high but for some reason, the confidence of the pose estimation network is super low. Leading to misidentified animals (which is probably also happening when predicting using the SLEAP API).
- [ ] It is still unclear how this affects identity. In Bonsai, we saw that when the pose estimation has high confidence so does the identity and vice-versa.

**Comment by lochhh** [comment]
With the recently acquired AEON3B dataset (camera focus fix, related to #168), the network was able to correctly identify animals in 99% of the hold-out test set which consists of single-animal frames (see [above](https://github.com/SainsburyWellcomeCentre/aeon_experiments/discussions/141#discussioncomment-5227713)). 

In multi-animal frames, there are still missing and swapped identities (as illustrated by the broken lines and large position changes in the figure below), especially when animals cross or when they are close to each other:
  - track occupancy: [0.88 0.91]
  - frames with <2 tracks: 0.21 ( 26127 / 126000 )
  
![output_AEON3B_NTP_TP_2](https://user-images.githubusercontent.com/14955489/228618310-28e9664a-8037-46b1-8461-482640fa2adf.png)

To do:
- [ ] Manually curate and label these 'problematic' frames
  - Using the raw pos data, extract frames with area > x (when animals cross)
  - Using SLEAP predictions, extract frames with fewer than 2 instances
  - Using SLEAP predictions, extract frames with large position changes (potential ID switches)
- [ ] Evaluate performance on multi-animal frames on a rolling basis
- [ ] Train network for animals with patterned tails  
- [ ] Train network for identifying animals in nest

**Reply by bruno-f-cruz** [reply]
Something that occurred to me. If you take a rolling mode filter of N samples over identity, how well do you do? Start with say 10 frames.

**Comment by lorycalca** [comment]
We have done an additional tail painting test with 3 mice as follow:

- NO PAINT  - mouse BAA-1102808
- PATTERN 1 - 1 BAND - mouse BAA-1102811 (test done in 'maintanance mode' by mistake)
- PATTERN 2 - 2 BANDS - mouse BAA-1102810

We have done solo tests and a test with all of them in the arena (B2). This was done on 29/03/2023 starting from about 10.40am. 
Basic annotations are also written.

**Reply by lochhh** [reply]
**training set:**
3346 single animal frames
98 multi-animal frames (when 2 or more animals fuse)

**test set:**
406 single
13 multi-animal frames (when 2 or more animals fuse)
Tracks identified: 439
Tracks correctly identified: 425
Total tracks: 445
ID accuracy: 0.9681093394077449

**multi-animal video:**
number of predicted instances: 0.8323592085235921 ( 273430 / 328500 )
frames with predicted instances: 1.0 ( 109500 / 109500 )
track occupancy: [0.86338813 0.86327854 0.77041096]
frames with < 3 tracks: 0.47062100456621003 ( 51533 / 109500 )

**Comment by jkbhagatio** [comment]
Closing this as we are no longer running these tail-paint sessions, and have now moved on to running tattoed-tail sessions (#260)
